packages/nimble/R/BUGS_modelDef.R

```diff
@@ -199,6 +200,58 @@ codeProcessIfThenElse <- function(code, constants, envir = parent.frame()) {
         return(code)
 }
 
+## This function recurses through a block of code and expands any submodels
+codeProcessModelMacros <- function(code, envir = parent.frame()) {
+    codeLength <- length(code)
+    ## First check if this is the start of a curly-bracketed block
+    if(code[[1]] == '{') {
+        if(codeLength > 1)
+            ## Recurse on each line
+            for(i in 2:codeLength)
+                code[[i]] <- codeProcessModelMacros(code[[i]], envir)
+        return(code)
+    }
+    ## If this is a for loop, recurse on the body of the loop
+    if(code[[1]] == 'for') {
+        code[[4]] <- codeProcessModelMacros(code[[4]], envir)
+        return(code)
+    }
+    ## Check if this line invokes a submodel.
+    ## This can be done in two ways:
+    ## (i) node1 [<- | ~] <macro name>(...)
+    ## or
+    ## (ii) <macro name>(args)
+    ##
+    ## The first version is more BUGS-like.
+    ## The second version allows more full control.
+
+    ## Check for first version:
+    possibleMacroName <- deparse(code[[1]])
+    if(possibleMacroName %in% c('<-', '~')) {
+        possibleMacroName <- deparse(code[[3]][[1]])
+    }
+    if(exists(possibleMacroName)) { ## may need to provide an envir argument
+        possibleMacro <- get(possibleMacroName) ## ditto
+        if(inherits(possibleMacro, "model_macro")) {
+            expandedInfo <- possibleMacro$process(code)
+            if(!is.list(expandedInfo))
+                stop(paste0("submodel ",
+                            possibleMacroName,
+                            " should return a list with an element named ",
+                            "'code'.  It did not return a list."))
+            if(!is.call(expandedInfo[['code']]))
+                stop(paste0("submodel ",
+                            possibleMacroName,
+                            " should return a list with an element named ",
+                            "'code' that is a call."))
+            ## possibly extract other content in the future
+            recursedNewCode <- codeProcessModelMacros(expandedInfo$code, envir)
```

Comment:
It seems a little more robust to simply assign `code <- codeProcessModelMacros(...)` here, rather than `return()` early. That way if someone adds postprocessing at the end of this function, it will apply to both code paths.